### PR TITLE
designate: Update ns_records with all nameservers (SOC-9636)

### PR DESCRIPTION
**Pull request by ritesh216**
Added FQDN for all nodes running dns-server.

### CODE REVIEW

**Comment by matthewoliver** [comment]
Ritesh is off having a baby, I'll take over the ticket.

Yeah have some new code to use the public-* hostnames. Just want to test
it, then will push.

File: chef/cookbooks/designate/recipes/mdns.rb [hunk 1]
```diff
@@ -34,12 +35,21 @@
 # with every delete/create cycle of proposal. This might mess
 # up the designate configuration. So the advantage of having
 # non-hardcoded is high enough
+
+# When available, use the public FQDN for each node.
+ns_records = dns_all.map { |dnss|
```
[the remark this answers is not in the document]
Comment (matthewoliver) [reply chef/cookbooks/designate/recipes/mdns.rb]: ok ok hound, I'll change it to 'do' and 'end'

**Comment by matthewoliver** [comment]
The downside to this approach is that in my test crowbar env, the pools list is now a mix because not all DNS nodes are connected to the Public network:

`---
- name: default-bind
  description: Default BIND9 Pool
  id: 794ccc2c-d751-44fe-b57f-8894c9f5c842
  attributes: {}
  ns_records:
  - hostname: public-d52-54-77-77-01-01.virtual.cloud.suse.de.
    priority: 1
  - hostname: d52-54-77-77-01-02.virtual.cloud.suse.de.
    priority: 1
  - hostname: crowbar.virtual.cloud.suse.de.
    priority: 1
  nameservers:
`

**Comment by matthewoliver** [comment]
Hrm, the other issue with listening on the public FQDN is we've only setup Bind in crowbar to listen on the admin ips.. so this actually wont work.
So really what ritesh had originally is probably more correct.

Or do we need to start looking as a split DNS setup in crowbar, with separating internal FQDNs with external public FQDN configurations.

**Comment by matthewoliver** [comment]
If we wanted to enable crowbar bind services to listen on the public network, we'd need a PR to crowbar-core with something like: http://pastebin.nue.suse.com/21306

Noting the ipv6 address would need to be plumbed through too, but this is just an example diff.
The enable_designate there is also only true currently when the server is the master.

**Comment by s-t-e-v-e-n-k** [comment]
We explicitly don't want that BIND listening on the public interface.

**Comment by matthewoliver** [comment]
Yeah, then I don't think we want to use the public FQDNs for the nameservers in the pools.crowbar.yaml. Or am I missing something obvious?

The ns_servers in the pools.yaml, according to designate is where it reads to confirms a change exists. Ie writes to the target and checks the ns_servers. So needs to be a slave or something. Which means it can talk to bind there right? 
So based on that, I'd say Ritish's original version _was_ correct

Though designate gives the users an API to manage DNS, so I assume that means the op/user wants access to the nameservers too. They can just create there own pool of name servers.

**Comment by matthewoliver** [comment]
So it comes down to the definition of public.  Ie not on the internet maybe, but public, ie outside the cloud so instances can access it... maybe?

Can we have more then 1 target in a pool. A target for just designate users and not one for crowbar... but I guess that's what multiple pools is all about.

**Comment by matthewoliver** [comment]
OK so we had a discussion on this during our stand up. If designate is designed to allow users to have DNSaaS, then the operator needs to provide a public accessible pool of DNS servers for designate to talk to.

The DNS servers we deploy with crowbar are really meant for the cloud and as such have only been setup, purposely, to only listen on the admin network. So any machine that is apart of the fabric of the cloud has access to it.. but not tenants and their instances.
Getting the crowbar DNS nodes to listen on the public network would be consisdered a security problem.

I guess in theory we could have an internal zones listening on the admin network and a public zones listening on the public network, ie split zone dns.. but again that doesn't exist atm and could end up being a can of worms when it comes to security.

**Comment by matthewoliver** [comment]
OK so looking at another card lower in the backlog (9635) we will get our crowbar DNS servers to listen on public. So have tidied up the patch to use the barclamp helper. This patch will use the public FQDN if the node is attached to the public network.

But really this patch will need to wait for a patch to crowbar-core coming out of 9635. When I've done it, I'll link it here.

**Comment by matthewoliver** [comment]
The first version of the crowbar-core patch has been created. Currently it's _very_ wip. But the PR link is https://github.com/crowbar/crowbar-core/pull/1887

**Comment by matthewoliver** [comment]
The other PR now works and ready for review. Which brings me to a question about this one. Should we only every use the public nameservers? Or as the patch currents implements it'll fall back to the admin FQDN. The designate service will be able to talk to them, and in theory so long as at least 1 is accessible via the public network, so can tenants.

**Comment by matthewoliver** [comment]
Now that's a good question :)

**Pull request merged**